### ISSUE TITLE
[longterm] Need to replace memory mapping for files

### ISSUE THREAD

**Issue by mrpapersonic** [schismtracker/schismtracker#521]
Memory mapping files kind of... sucks. The best way I can think of to do this is to retrofit the slurp code into being a glorified `FILE *`. In addition, I'd like to refactor the init functions to modify a passed pointer instead of forcing heap allocation.

**Comment by mrpapersonic** [comment]
slowly picking at this, the plan so far is
1. replace all of the read info, load sample, and load instrument function definitions with a `slurp_t *` instead of `uint8_t *, size_t`. simple readers can easily be adapted here, but more complex ones will have to be hooked to `fp->data` and `fp->length`.
2. replace the remaining memory logic (the most complex stuff is in the dsm and aiff loaders) with calling slurp functions
3. then adapt all of the existing slurp code to use an internal `FILE *`

Currently I'm partly through step 1. The IT loaders are particularly dependent on memory access because the code had to be shared between the load module/sample/instrument code, but adapting them shouldn't be that difficult.

**Comment by mrpapersonic** [comment]
Okay, *removing* memory mapping might make a performance toll. I'm instead gonna use `FILE *` as a *fallback* for when `mmap` or `MapViewOfFile` doesn't work or they don't exist (which will probably happen on some weird esoteric platforms).

**Comment by sagamusix** [comment]
You'll probably want to have at least some basic caching in slurp (4KB blocks or so) in order to keep `FILE *` access somewhat performant. FYI, OpenMPT went through all of those steps as well - starting with direct poking at memory-mapped files, then rewriting all loaders to use an abstraction layer (sometimes backporting Schism's slurp-based loaders where it made sense), then augmenting this abstraction layer with the possibility of using stuff like `fstream` as the data provider, but also adding an internal cache to avoid going through other slow abstractions for reading a single byte.

**Comment by mrpapersonic** [comment]
I'm not really worried about adding a cache buffer for now, since I only plan on adding a stdio backend, which is already buffered in the first place. We can also specifically request a buffer size of 4 KiB through `setvbuf(fp, NULL, _IOFBF, 4096ULL);`.

This could even be *faster* than memory mapping, in particular for the info readers since most of those only need the first 4 KiB anyway to recognize a legitimate file...

I'm guessing the main pain point would be figuring out an implementation of `IMFByteStream` for Media Foundation, which I believe should, among the other benefits, bring Media Foundation support to the grand total of 3 people that still use Vista for some reason.

**Comment by mrpapersonic** [comment]
This is now implemented on the master branch. To reiterate, memory mapping wasn't removed, but the stupid approach of reading the entire file into memory was canned in favor of using the C stdio `FILE *`.

Implementing IMFByteStream wasn't *that* difficult; in fact, the implementation worked first try once I got everything compiling correctly. The only things that were broken were unrelated changes in behavior that I did...

What's annoying now though, is that Media Foundation wants to say that every .exe file is an MPEG-1 file (???). Maybe this was an issue before and I just never noticed it, but I'm considering just filtering out any file that starts with `MZ`, since it's really annoying.
